parachain/pallets/bridge/src/lib.rs

```diff
@@ -24,32 +26,75 @@ use frame_system::{self as system, ensure_signed};
 use sp_std::prelude::*;
 use sp_core::H160;
 
+use sp_runtime::RuntimeDebug;
+
+use codec::{Encode, Decode}
+
 use artemis_core::{AppId, Application, Message, Verifier};
 
+use primitives::{self, ChannelId, MessageNonce};
+
+pub mod primitives;
+pub mod inbound_channel;
+
+pub trait ApplicationRegistry {
+
+	fn iter() -> Iterator<Application>;
+}
+
 pub trait Config: system::Config {
 	type Event: From<Event> + Into<<Self as system::Config>::Event>;
 
 	/// The verifier module responsible for verifying submitted messages.
 	type Verifier: Verifier<<Self as system::Config>::AccountId>;
 
-	/// ETH Application
-	type AppETH: Application;
+	type ApplicationRegistry: ApplicationRegistry;
+}
+// Storage for inbound channels
+//
+// Adapted from parity-bridges-common (modules/message-lane/lib.rs)
+struct RuntimeInboundChannelStorage<T: Config> {
+	cached_data: RefCell<Option<InboundChannelData>>,
+}
 
-	/// ERC20 Application
-	type AppERC20: Application;
+impl<T: Config> RuntimeInboundChannelStorage {
+	fn data(&self) -> InboundChannelData {
+		match self.cached_data.clone().into_inner() {
+			Some(data) => data,
+			None => {
+				let data = InboundChannelData::<T>::get(&self.lane_id);
+				*self.cached_data.try_borrow_mut().expect(
+					"we're in the single-threaded environment;\
+						we have no recursive borrows; qed",
+				) = Some(data.clone());
+				data
+			}
+		}
+	}
+
+	fn set_data(&mut self, data: InboundLaneData<T::InboundRelayer>) {
+		*self.cached_data.try_borrow_mut().expect(
+			"we're in the single-threaded environment;\
+				we have no recursive borrows; qed",
+		) = Some(data.clone());
+		InboundChannelData::<T>::set(data)
+	}
 }
 
 decl_storage! {
 	trait Store for Module<T: Config> as BridgeModule {
-
+		InboundChannelData: primitives::InboundChannelData;
 	}
 }
 
-decl_event!(
+decl_event! {
     /// Events for the Bridge module.
 	pub enum Event {
+
+		/// Message accepted for delivery to Ethereum
+		MessageAccepted(ChannelId, Option<MessageNonce>)
```

Comment:
Parity-bridges-common also emits a `MessageDelivered` if the Ethereum side actually got the message. Pretty neat!

Reply:
you mean on the ethereum smart contract? or where? link?

Reply:
Here we go: https://github.com/paritytech/parity-bridges-common/blob/master/modules/message-lane/src/lib.rs#L212

Reply:
oh, yeh that's with a delivery confirmation flow too - our design doesn't rely on this or want/need this

Reply:
In theory could be implemented like this: We just need the ReceiveChannel on the Ethereum side ask its SendChannel sibling to send some kind of "message delivered" control message across the bridge.

Of course this is a nice-to-have and can wait till after we've delivered.

Reply:
yes but it's not useful/valuable at all in our current design, so not even a nice-to-have.

for alternative designs/setups, yes it could be added as such

Reply:
Have removed `MessageAccepted`